crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -1862,23 +1862,24 @@ macro_rules! impl_anytuple_fetch {
             }
 
             fn update_component_access(state: &Self::State, _access: &mut FilteredAccess<ComponentId>) {
-                let ($($name,)*) = state;
+                // update the access (add the read/writes)
+                <($(Option<$name>,)*)>::update_component_access(state, _access);
 
+                // update the filters (Or<(With<$name>,)>)
+                let ($($name,)*) = state;
                 let mut _new_access = _access.clone();
                 let mut _not_first = false;
                 $(
                     if _not_first {
                         let mut intermediate = _access.clone();
                         $name::update_component_access($name, &mut intermediate);
```

Comment:
After the `<($(Option<$name>,)*)>::update_component_access(state, _access);` the variable `_access` contains the access to the various components, but this makes the following `$name::update_component_access($name, &mut intermediate);` fail because they see the access already registered. 

A solution might be to clone `_access` before the initial `<($(Option<$name>,)*)>::update_component_access(state, _access);`, then use the cloned one to create the various `intermediate`s.

Reply:
You're right, I was looking at this independently.
I think I found a better solution:
- we don't need to call `$name::update_component_access($name, &mut _access.clone());` since we're calling it just to update the filters, not the access. Instead we can just call `$name::update_component_access($name, &mut FilteredAccess:default());`
- the weird `_not_first` thing is because the `filter_sets` are initialized with an empty filter set. So we have a special case for the first element because if we just called `append_or` we would get the list of filters `Or<((), With<A>)>`, whereas we want `Or<(With<A>)>`

(we can also get rid of the `_not_first` thing in the `Or` filter, but I'll do it in another PR)

Reply:
Nevermind it was more complicated than expected, I did what you suggested @SkiFire13